### PR TITLE
small vector

### PR DESCRIPTION
**Pull request by havardpe**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@arnej27959 please review
@geirst FYI

### CODE REVIEW

File: vespalib/src/vespa/vespalib/util/small_vector.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
```
Comment (arnej27959) [inline vespalib/src/vespa/vespalib/util/small_vector.h:2]: missing #pragma once

**Review by arnej27959 — APPROVED** [review]
looks good to me.
i think we want some way to construct with size, for example
SmallVector(size_t initial_size, const T& value)

**Comment by havardpe** [comment]
@arnej27959 PTAL

**Comment by vekterli** [comment]
Re: the LLVM version of this, should we also consider having a `SmallVectorImpl`/`SmallVectorBase` concept that type-erases away the `N`?

**Review by arnej27959 — APPROVED** [review]
👍

**Comment by arnej27959** [comment]
after trying to use it some random places, I think we rather want conversion to and from ConstArrayRef.

**Comment by havardpe** [comment]
> Re: the LLVM version of this, should we also consider having a `SmallVectorImpl`/`SmallVectorBase` concept that type-erases away the `N`?

Maybe in the future. Erasing the N is only really useful if we want to use it for APIs that modify our vector or we want to use it with different N for the same T. It also complicate things like knowing when the elements are externally allocated and reverting a vector back to its initial state.

**Comment by havardpe** [comment]
> after trying to use it some random places, I think we rather want conversion to and from ConstArrayRef.

Lets talk about the convenient uses and I will make a follow-up PR for those

**Pull request merged**